app/Http/Controllers/ApiController.php

```diff
@@ -102,6 +104,69 @@ public function update(Request $request)
     }
     
     /**
+     * Updates the sensors of a device.
+     *
+     * @param  Request  $request
+     * @return Response
+     */
+    public function sensor(Request $request)
+    {
+        // Validate the request
+        $validator = Validator::make($request->all(), [
+            'uuid'          => 'required|string|max:255|exists:devices,uuid',
+            'token'         => 'required|string|max:60',
+            'sensor_data.*.name'   => 'required',
+            'sensor_data.*.type'   => 'required',
+            'sensor_data.*.value'  => 'required',
+        ]);
+        
+        // If validation fails, send the validation error back with status 400.
+        if ($validator->fails()) {
+            return response()->json(['data' => $validator->errors()], 400);
+        }
+        
+        // Get the device record.
+        $device = Device::getDeviceByUUID($request->input('uuid'));
+        
+        // If token doesnt match then send 401 unauthorized.
+        if ($request->input('token') != $device->token) {
+            return response()->json(['data' => 'Bad token.'], 401);
+        }
+        
+        // Update the device.
+// 		"sensor_data": {
+// 			{ "name": "cpu", "type": "cpu_temperature", "value": cpu_temp() },
+// 			{ "name": "temperature", "type": "temperature", "value": temperature() },
+// 			{ "name": "humidity", "type": "humidity", "value": humidity() },
+// 			{ "name": "moisture_01", "type": "moisture", "value": 0.00 },
+// 			{ "name": "moisture_02", "type": "moisture", "value": 0.00 },
+// 			{ "name": "light_in", "type": "light", "value": 0.00 },
+// 			{ "name": "light_out", "type": "light", "value": 0.00 }
+// 		}
+        $sensor_datas = $request->input('sensor_data');
+        foreach ($sensor_datas as $sensor_data) {
```

Comment:
The expression $sensor_datas of type string|array is not guaranteed to be traversable. How about adding an additional type check?

Reply:
If you validate the array to be a min of 1 I think you should be good. I haven't tested but something like: 
```php
"sensor_data" => 'required|array|min:1',
```
Then do you regular validation on the params like
```php
'sensor_data.*.name'   => 'required|string|max:50',
'sensor_data.*.type'   => 'required|string|max:50',
'sensor_data.*.value'   => 'required|string|max:50',
```

Reply:
False positive, Validation requires that at least the first index of the string|array is populated.